### PR TITLE
Fix Phobos issue 15480

### PR DESCRIPTION
**Pull request by tsbockman**
This fixes [Phobos issue #15480: std.algorithm.iteration.map not accepting multiple lambdas](https://issues.dlang.org/show_bug.cgi?id=15480).

The solution is rather hackish, due to the need to work around [DMD issue #5710: cannot use delegates as parameters to non-global template](https://issues.dlang.org/show_bug.cgi?id=5710).

This PR is intended to supersede [PR #3963](https://github.com/D-Programming-Language/phobos/pull/3963) by @tirkarthi .

### CODE REVIEW

**Comment by tsbockman** [comment]
@tirkarthi 
> Also I think the unit tests can be moved to https://github.com/tsbockman/phobos/blob/be7c55fbea7f37987e9f965f968c54b4a3f7342b/std/algorithm/iteration.d#L753 and the test static assert(!__traits(compiles, map!(c => voidFun(c))([1, 2, 3]))); can be added for the unittest with void lambdas.

I have updated the unit tests.

**Comment by tirkarthi** [comment]
@tsbockman Thanks. I closed my previous PR with reference to the current PR. Hope someone from the community reviews it.

**Comment by tsbockman** [comment]
@klickverbot 
> I'd consider removing the HACK comment though, as this implementation is probably better in terms of compile time performance anyway (the loop is not a runtime loop, so…).

I guess `HACK` is too strong of a label; I was thinking of an earlier, wonkier version of this when I wrote that.

But, this way still feels wrong to me. Generally speaking I try to write my compile-time logic in a way that is guaranteed not to emit any runtime code - either by sticking it in a `template`, or an `enum _ = function() { }();` block.

While it's true that `foreach` over an `AliasSeq` is supposed to be a compile-time construct, it *looks* like a runtime one unless you take the time to track down the definition of the collection being iterated over. (Admittedly, that's trivial in this case.)

I wish they'd called it `static foreach`, to prevent confusion with regular runtime `foreach`.

**Comment by tsbockman** [comment]
@klickverbot 
> Auto-merge toggled off

Is there a problem? The last update I did was just removing the `HACK` label, as you suggested.

**Comment by MartinNowak** [comment]
Seems like this introduced a regression.
```d
import std.algorithm, std.string;

auto foo(string[] args)
{
    return args.map!strip;
}
```
Weird enough, it compiles when I import `std.algorithm.iteration` instead.

**Comment by tsbockman** [comment]
@MartinNowak 
> Weird enough, it compiles when I import `std.algorithm.iteration` instead.

You have an import conflict between `std.algorithm.mutation.strip` and `std.string.strip`.

I don't know if your example should ever have compiled, but if so I would check into recent DMD work, instead.

**Comment by CyberShadow** [comment]
I have hit the same regression.

> You have an import conflict between `std.algorithm.mutation.strip` and `std.string.strip`.
> 
> I don't know if your example should ever have compiled, but if so I would check into recent DMD work, instead.

No, I think you're wrong. There is no conflict, only an overload set.

The error message is:

`variable std.algorithm.iteration.map!(strip).map!(string[]).map.f variables cannot be of type void`

It happens on this line:

`foreach(f; _funs)`

I'm not yet sure what's going on, but I think it's expanding `_funs` and attempting to evaluate it or something.

There is clearly a regression here.

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -748,6 +748,10 @@ unittest
     static assert(!__traits(compiles, map!(voidFun, voidFun)([1])));
     static assert(!__traits(compiles, map!(nonvoidFun, voidFun)([1])));
     static assert(!__traits(compiles, map!(voidFun, nonvoidFun)([1])));
+    static assert(!__traits(compiles, map!(a => voidFun(a))([1])));
+
+    // Phobos issue #15480
+    auto dd = map!(z => z * z, c => c * c * c)([ 1, 2, 3, 4 ]);
```
Comment (CyberShadow) [inline std/algorithm/iteration.d]: No check for the returned values?

**Comment by CyberShadow** [comment]
So this change brought up a latent DMD bug as a regression. I filed the DMD bug: https://issues.dlang.org/show_bug.cgi?id=15777

This should still be worked around in Phobos. E.g.:

```
template map(fun...) if (fun.length >= 1)
{
    auto map(Range)(Range r) if (isInputRange!(Unqual!Range))
    {
        import std.meta : AliasSeq, staticMap;

        alias RE = ElementType!(Range);
        static if (fun.length > 1)
        {
            import std.functional : adjoin;
            import std.meta : staticIndexOf;

            alias _funs = staticMap!(unaryFun, fun);
            alias _fun = adjoin!_funs;

	        // Once DMD issue #5710 is fixed, this validation loop can be moved into a template.
			foreach(f; _funs)
			{
				static assert(!is(typeof(f(RE.init)) == void),
					"Mapping function(s) must not return void: " ~ _funs.stringof);
			}
        }
        else
        {
            alias _fun = unaryFun!(fun[0]);
            alias _funs = AliasSeq!(_fun);

			// Do the validation separately for single parameters due to DMD issue #3969.
			static assert(!is(typeof(_fun(RE.init)) == void),
				"Mapping function(s) must not return void: " ~ _funs.stringof);
        }

        return MapResult!(_fun, Range)(r);
    }
}
```

@tsbockman Will you do the PR or should I?

**Comment by tsbockman** [comment]
> @tsbockman Will you do the PR or should I?

I made it: [Phobos PR #4065](https://github.com/D-Programming-Language/phobos/pull/4065).

**Comment by CyberShadow** [comment]
Thanks!

**Comment by CyberShadow** [comment]
> So this change brought up a latent DMD bug as a regression.

Another one:

https://issues.dlang.org/show_bug.cgi?id=17155

**Pull request merged**